File: 216.combination-sum-iii.java
/*
 * @lc app=leetcode id=216 lang=java
 *
 * [216] Combination Sum III
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

class Solution {
    private List<Integer> path;
    private List<List<Integer>> result;
    public List<List<Integer>> combinationSum3(int k, int n) {
        path = new ArrayList<>();
        result = new ArrayList<>();
        backtracking(k, n, 1);
        return result;
    }
    private void backtracking(int k, int n, int startIndex) {
        if (path.size() == k) {
            if (path.stream().mapToInt(x -> x).sum() == n) {
                result.add(new ArrayList<>(path));
            }
            return;
        }
        if (startIndex + (k - path.size()) > 10) return;
        for (int i = startIndex; i != 10; ++i) {
            path.add(i);
            backtracking(k, n, i + 1);
            path.removeLast();
        }
    }
}
// @lc code=end

File: 216.combination-sum-iii.cpp
/*
 * @lc app=leetcode id=216 lang=cpp
 *
 * [216] Combination Sum III
 */

// @lc code=start

#include <vector>
#include <numeric>

using namespace std;

class Solution {
public:

    vector<vector<int>> results;
    vector<int> path;
    int sum {};

    vector<vector<int>> combinationSum3(int k, int n) {
        backtracking(k, n, 1);
        return results;
    }

    void backtracking(int k, int n, int startIndex) {
        if (k > n) return;
        if (sum > n) return;
        if (path.size() == k) {
            if (sum == n) {
                results.push_back(path);
            }
            return;
        }
        for (int ix = startIndex; ix != 10; ++ix) {
            path.push_back(ix);
            sum += ix;
            backtracking(k, n, ix + 1);
            sum -= ix;
            path.pop_back();
        }
    }
};
// @lc code=end
